File: java/tsfile/src/main/java/org/apache/tsfile/read/v4/TsFileTreeReader.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.tsfile.read.v4;

import org.apache.tsfile.annotations.TreeModel;
import org.apache.tsfile.annotations.TsFileApi;
import org.apache.tsfile.file.metadata.StringArrayDeviceID;
import org.apache.tsfile.read.TsFileReader;
import org.apache.tsfile.read.common.Path;
import org.apache.tsfile.read.expression.IExpression;
import org.apache.tsfile.read.expression.QueryExpression;
import org.apache.tsfile.read.expression.impl.GlobalTimeExpression;
import org.apache.tsfile.read.filter.operator.TimeFilterOperators;
import org.apache.tsfile.read.query.dataset.QueryDataSet;
import org.apache.tsfile.read.query.dataset.ResultSet;
import org.apache.tsfile.read.query.dataset.TreeResultSet;
import org.apache.tsfile.write.schema.MeasurementSchema;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * TsFileTreeReader is an implementation of ITsFileTreeReader that provides query capability for
 * TsFile with a tree-style interface. It internally wraps {@link TsFileReader}.
 */
public class TsFileTreeReader implements ITsFileTreeReader {
  private final TsFileReader tsfileReader;

  /**
   * Construct a TsFileTreeReader with the given TsFile.
   *
   * @param file the TsFile to read
   * @throws IOException if an I/O error occurs during file opening
   */
  @TsFileApi
  @TreeModel
  public TsFileTreeReader(File file) throws IOException {
    tsfileReader = new TsFileReader(file);
  }

  /**
   * Execute a query on the given devices and measurements within the specified time range.
   *
   * @param deviceIds list of device IDs to query
   * @param measurementNames list of measurement names to query
   * @param startTime query start timestamp (inclusive)
   * @param endTime query end timestamp (inclusive)
   * @return a {@link ResultSet} containing the query results
   * @throws IOException if an I/O error occurs during query execution
   */
  @TsFileApi
  @TreeModel
  @Override
  public ResultSet query(
      List<String> deviceIds, List<String> measurementNames, long startTime, long endTime)
      throws IOException {
    List<Path> paths = new ArrayList<>();
    for (String deviceId : deviceIds) {
      for (String measurementName : measurementNames) {
        paths.add(new Path(deviceId, measurementName, true));
      }
    }
    IExpression expression =
        new GlobalTimeExpression(new TimeFilterOperators.TimeBetweenAnd(startTime, endTime));
    QueryExpression queryExpression = QueryExpression.create(paths, expression);
    QueryDataSet queryDataSet = tsfileReader.query(queryExpression);
    return new TreeResultSet(queryDataSet, deviceIds, measurementNames);
  }

  /**
   * Get all device IDs existing in the TsFile.
   *
   * @return list of all device IDs
   * @throws IOException if an I/O error occurs during metadata fetching
   */
  @TsFileApi
  @TreeModel
  @Override
  public List<String> getAllDeviceIds() throws IOException {
    return tsfileReader.getAllDeviceIds();
  }

  /**
   * Get the measurement schema of the given device.
   *
   * @param deviceId the target device ID
   * @return list of {@link MeasurementSchema} for the device
   * @throws IOException if an I/O error occurs during schema fetching
   */
  @TsFileApi
  @TreeModel
  @Override
  public List<MeasurementSchema> getDeviceSchema(String deviceId) throws IOException {
    return tsfileReader.getMeasurement(new StringArrayDeviceID(deviceId));
  }

  /**
   * Close the TsFileTreeReader and release resources.
   *
   * @throws IOException if an I/O error occurs during closing
   */
  @TsFileApi
  @TreeModel
  @Override
  public void close() throws IOException {
    tsfileReader.close();
  }
}

File: cpp/src/reader/tsfile_tree_reader.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * License); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "reader/tsfile_tree_reader.h"

namespace storage {

TsFileTreeReader::TsFileTreeReader() {
    tsfile_reader_ = std::make_shared<TsFileReader>();
}

TsFileTreeReader::~TsFileTreeReader() = default;

int TsFileTreeReader::open(const std::string& file_path) {
    return tsfile_reader_->open(file_path);
}

int TsFileTreeReader::close() { return tsfile_reader_->close(); }

int TsFileTreeReader::query(const std::vector<std::string>& device_ids,
                            const std::vector<std::string>& measurement_names,
                            int64_t start_time, int64_t end_time,
                            ResultSet*& result_set) {
    std::vector<std::string> path_list;
    for (auto& device_id : device_ids) {
        for (auto& measurement : measurement_names) {
            path_list.emplace_back(device_id + PATH_SEPARATOR_CHAR +
                                   measurement);
        }
    }
    return tsfile_reader_->query(path_list, start_time, end_time, result_set);
}

void TsFileTreeReader::destroy_query_data_set(ResultSet* qds) {
    tsfile_reader_->destroy_query_data_set(qds);
}

std::vector<MeasurementSchema> TsFileTreeReader::get_device_schema(
    const std::string& device_id) {
    std::vector<MeasurementSchema> schemas;
    tsfile_reader_->get_timeseries_schema(
        std::make_shared<StringArrayDeviceID>(device_id), schemas);
    return schemas;
}

std::vector<std::string> TsFileTreeReader::get_all_device_ids() {
    std::vector<std::string> ret_device_ids;
    auto device_ids = tsfile_reader_->get_all_device_ids();
    for (auto device_id : device_ids) {
        ret_device_ids.emplace_back(device_id->get_device_name());
    }
    return ret_device_ids;
}

std::vector<std::shared_ptr<IDeviceID>> TsFileTreeReader::get_all_devices() {
    return tsfile_reader_->get_all_devices();
}

DeviceTimeseriesMetadataMap TsFileTreeReader::get_timeseries_metadata(
    const std::vector<std::shared_ptr<IDeviceID>>& device_ids) {
    return tsfile_reader_->get_timeseries_metadata(device_ids);
}

DeviceTimeseriesMetadataMap TsFileTreeReader::get_timeseries_metadata() {
    return tsfile_reader_->get_timeseries_metadata();
}

}  // namespace storage
